compiler/rustc_ast/src/token.rs

```diff
@@ -216,6 +216,9 @@ pub fn ident_can_begin_expr(name: Symbol, span: Span, is_raw: IdentIsRaw) -> boo
             kw::Static,
         ]
         .contains(&name)
+        // Please note the order of the check here, the edition is not check on every
+        // token by only on the possible one.
+        || (kw::Underscore == name && span.edition().at_least_rust_2024())
```

Comment:
I am not certain if this is the best approach. I will need to look at the callsites that use this function.

Reply:
Yeah sorry, I noted that this is causing a compilation error, while in the meeting with @eholk today I propose that I look more at https://github.com/rust-lang/rust/pull/126700 and see if there is another way to do it.

Reply:
Assuming this comment is about this code: (The PR has changed some since this comment was posted and I'm not sure how good GitHub is at tracking comments across revisions)
```rust
        // Please note the order of the check here, the edition is not check on every
        // token by only on the possible one.
        || (kw::Underscore == name && span.edition().at_least_rust_2024())
```

This seems like a change that likely has broader impacts than we expect, as it causes `_` to be treated as an expression _everywhere_. My understanding is that `_` is only a valid expression in certain cases, but for the purposes of macros we were going to consider it always an expression. So the cases where the rest of the compiler handles `_` as an expression are likely handled already through a different mechanism.

I'm guessing we want something that's scoped closer to this line over at line 151 of nonterminal.rs:

```rust
            NonterminalKind::Expr(_) => NtExpr(self.parse_expr_force_collect()?),
```

Reply:
You are right @eholk, the `ident_can_begin_expr` will impact more parts of the compiler. 

Thanks for catching it. However, I tried to understand the `parse_expr_force_collect` but was a little bit too messy for me, and while I was reading the code I found out that we should extend the `token.can_begin_expr()` just for the `NonterminalKind::Expr(Expr)`.

I pushed the code, let me know if I am going off the way for implementing a minimal solution that works for macros declaration.